tez-runtime-library/src/main/java/org/apache/tez/dag/library/vertexmanager/VertexManagerWithConcurrentInput.java

```diff
@@ -0,0 +1,242 @@
+    }
+    managedTasks = getContext().getVertexNumTasks(getContext().getVertexName());
+    Map<String, EdgeProperty> edges = getContext().getInputVertexEdgeProperties();
+    for (Map.Entry<String, EdgeProperty> entry : edges.entrySet()) {
+      if (!CONCURRENT.equals(entry.getValue().getSchedulingType())) {
+        throw new TezUncheckedException("All input edges to vertex " + vertexName +
+            "  must be CONCURRENT.");
+      }
+      String srcVertex = entry.getKey();
+      srcVerticesConfigured.put(srcVertex, false);
+      getContext().registerForVertexStateUpdates(srcVertex, EnumSet.of(VertexState.CONFIGURED));
+    }
+
+    try {
+      vertexConfig = TezUtils.createConfFromUserPayload(getContext().getUserPayload());
+    } catch (IOException e) {
+      throw new TezUncheckedException(e);
+    }
+    edgeTriggerType = ConcurrentEdgeTriggerType.valueOf(
+        vertexConfig.get(TezConfiguration.TEZ_CONCURRENT_EDGE_TRIGGER_TYPE,
+            TezConfiguration.TEZ_CONCURRENT_EDGE_TRIGGER_TYPE_DEFAULT));
+    if (!ConcurrentEdgeTriggerType.SOURCE_VERTEX_CONFIGURED.equals(edgeTriggerType)) {
+      // pending TEZ-3999
+      throw new TezUncheckedException("Only support SOURCE_VERTEX_CONFIGURED triggering type for now.");
+    }
+    LOG.info("VertexManagerWithConcurrentInput initialized with edgeTriggerType {}.", edgeTriggerType);
+
+    vertexName = getContext().getVertexName();
+    completedUpstreamTasks = 0;
+  }
+
+  @Override
+  public void onVertexStarted(List<TaskAttemptIdentifier> completions) {
+    onVertexStartedDone.set(true);
+    scheduleTasks();
+  }
+
+  @Override
+  public void onVertexStateUpdated(VertexStateUpdate stateUpdate) {
+    VertexState state = stateUpdate.getVertexState();
+    String fromVertex = stateUpdate.getVertexName();
+    if (!srcVerticesConfigured.containsKey(fromVertex)) {
+      throw new IllegalArgumentException("Not expecting state update from vertex:" +
+          fromVertex + " in vertex: " + this.vertexName);
+    }
+
+    if (!VertexState.CONFIGURED.equals(state)) {
+      throw new IllegalArgumentException("Received incorrect state notification : " +
+          state + " from vertex: " + fromVertex + " in vertex: " + this.vertexName);
+    }
+
+    LOG.info("Received configured notification: " + state + " for vertex: "
+        + fromVertex + " in vertex: " + this.vertexName);
+    srcVerticesConfigured.put(fromVertex, true);
+
+    // check for source vertices completely configured
+    boolean checkAllSrcVerticesConfigured = true;
+    for (Map.Entry<String, Boolean> entry : srcVerticesConfigured.entrySet()) {
+      if (!entry.getValue()) {
+        // vertex not configured
+        LOG.info("Waiting for vertex {} in vertex {} ", entry.getKey(), this.vertexName);
+        checkAllSrcVerticesConfigured = false;
+        break;
+      }
+    }
+    allSrcVerticesConfigured = checkAllSrcVerticesConfigured;
```

Comment:
Doing it like this makes me wonder if the tasks can be scheduled twice for example `onVertexStateUpdated` being called twice closely:

* Thread 1: calls onVertexStateUpdated
* Thread 2: calls onVertexStateUpdated
* Thread 1: at `allSrcVerticesConfigured = checkAllSrcVerticesConfigured;` is set to false
* Thread 2: at `allSrcVerticesConfigured = checkAllSrcVerticesConfigured;` is set to true now
* Thread 1: scheduleTasks schedules all the tasks because  `allSrcVerticesConfigured` is true
* Thread 2: scheduleTasks schedules all the tasks because  `allSrcVerticesConfigured` is true

I see the variable `tasksScheduled` is used, but I think it could still happen. Maybe with some `AtomicBoolean.getAndSet` we'd be safer.

Reply:
Sorry I meant `compareAndSet` to make sure `getContext().scheduleTasks(tasksToStart)` is not called twice. I think now it could still happen. So it would be something like
```java
if (tasksScheduled .compareAndSet(false, true)) {
     // schedule tasks
}
```

Reply:
i see what you meant. done

Reply:
What I'm saying is that two threads can still go pass `if (tasksScheduled.get()) {` when it's `false`, then `if (!canScheduleTasks()) {` would be true for both threads and it doesn't matter if you `tasksScheduled.compareAndSet` if you don't check the return value.

Reply:
I think the source of risk here is from the plugin interfaces are not synchronized. I have made all meaningful implementation of the plugin interfaces (onVertexStateUpdated, onSourceTaskCompleted, etc.) synchronized, following the common practice of other plugin implmentation (such as ShuffleVertexManagerBase).

We should be good now.

Reply:
I don't think it's necessary to synchronize all the methods and that would unnecessarily slow this down when an `AtomicBoolean` to guard `getContext().scheduleTasks(tasksToStart)` would be enough IMO. 
This does solve the bug though so I'd be happy to create a follow up ticket to address this and let this go in how it is in this PR. Maybe we can look into this again when we see the whole picture when the other tickets implemented.